File: src/problems/leetcode-cn/17. 电话号码的字母组合/solution/java/Solution.java
class Solution {
    static final Map<Character, char[]> DIC = Map.of(
        '2', new char[]{'a', 'b', 'c'},
        '3', new char[]{'d', 'e', 'f'},
        '4', new char[]{'g', 'h', 'i'},
        '5', new char[]{'j', 'k', 'l'},
        '6', new char[]{'m', 'n', 'o'},
        '7', new char[]{'p', 'q', 'r', 's'},
        '8', new char[]{'t', 'u', 'v'},
        '9', new char[]{'w', 'x', 'y', 'z'}
    );

    List<String> result = new LinkedList<>();

    public List<String> letterCombinations(String digits) {
        List<char[]> seq = new LinkedList<>();
        for (int i = 0; i < digits.length(); i++) {
            seq.add(DIC.get(digits.charAt(i)));
        }
        backtrack(seq, "", 0);
        return result;
    }

    void backtrack(List<char[]> arr, String selected, int p) {
        if (selected.length() == arr.size()) {
            result.add(selected);
            return;
        }
        char[] ac = arr.get(p);
        for (int i = 0; i < ac.length; i++) {
            backtrack(arr, selected + ac[i], p+1);
        }
    }
}

File: src/problems/leetcode-cn/17. 电话号码的字母组合/solution/go/solution.go
var dic = map[byte][]byte {
    '2': []byte{'a', 'b', 'c'},
    '3': []byte{'d', 'e', 'f'},
    '4': []byte{'g', 'h', 'i'},
    '5': []byte{'j', 'k', 'l'},
    '6': []byte{'m', 'n', 'o'},
    '7': []byte{'p', 'q', 'r', 's'},
    '8': []byte{'t', 'u', 'v'},
    '9': []byte{'w', 'x', 'y', 'z'},
}

func letterCombinations(digits string) []string {
    result, collection := []string{}, [][]byte{}

    var backtrack func(arr [][]byte, selected []byte, p int)
    backtrack = func(arr [][]byte, selected []byte, p int) {
        if p == len(arr) {
            result = append(result, string(selected))
            return
        }
        for i := 0; i < len(arr[p]); i++ {
            backtrack(arr, append(selected, arr[p][i]), p+1)
        }
    }

    for i := 0; i < len(digits); i++ {
        collection = append(collection, dic[digits[i]])
    }

    backtrack(collection, []byte{}, 0)

    return result
}